quipucords/api/reports/serializer.py

```diff
@@ -1,71 +1,54 @@
 )
 
-from api.models import (
-    Report,
-    ScanJob,
-    ScanTask,
-)
-from constants import MINIMUM_REPORT_VERSION, DataSources
-
-
-class SourceSerializer(Serializer):
-    """Serializer for 'details report' sources."""
+from api.models import ScanJob, ScanTask
+from api.reports.model import Report
+from api.source.serializer import InternalSourceSerializer
 
-    server_id = UUIDField()
-    report_version = CharField()
-    source_name = CharField()
-    source_type = ChoiceField(choices=DataSources.choices)
-    facts = ListField(child=DictField(), min_length=1)
-
-    def validate_report_version(self, value):
-        """Validate report version."""
-        regex_match = re.match(r"(\d+)\.(\d+)\.(\d+)\+\S", value)
-        if not regex_match:
-            raise ValidationError("Invalid report_version.")
-        source_version = tuple(map(int, regex_match.groups()))
-        if source_version < MINIMUM_REPORT_VERSION:
-            source_version_str = ".".join(map(str, source_version))
-            minimum_version_str = ".".join(map(str, MINIMUM_REPORT_VERSION))
-            raise ValidationError(
-                f"Source version ({source_version_str}) is below the minimum "
-                f"supported report version ({minimum_version_str})"
-            )
-        return value
-
-
-class ReportUploadSerializer(Serializer):
-    """Serializer for uploaded details reports."""
-
-    report_type = ChoiceField(choices=["details"])
-    sources = SourceSerializer(many=True)
+SCANJOB_QUERYSET = ScanJob.objects.filter(
+    # scan_type=ScanTask.SCAN_TYPE_INSPECT,
+    status=ScanTask.COMPLETED,
+)
 
-    def validate_sources(self, value):
-        """Validate sources."""
-        if len(value) == 0:
-            raise ValidationError("minimum length required for 'sources' is 1.")
-        return value
 
-    @transaction.atomic
-    def create(self, validated_data):
-        """Create a report filled with raw data and a fingerprint-type ScanJob."""
-        report = Report.objects.create()
-        # create a scan job and required scan tasks for a "upload" job
-        scan_job = ScanJob.objects.create(
-            scan_type=ScanTask.SCAN_TYPE_FINGERPRINT,
-            report=report,
-        )
-        scan_job.ingest_sources(validated_data["sources"])
-        return report
+class CreateReportSerializer(Serializer):
+    from_jobs = ListField(
+        child=PrimaryKeyRelatedField(queryset=SCANJOB_QUERYSET),
+        required=False,
+    )
+    from_reports = ListField(
+        child=PrimaryKeyRelatedField(
+            queryset=Report.objects.filter(scanjob__in=SCANJOB_QUERYSET)
+        ),
+        required=False,
+    )
+
+    def validate(self, attrs):
+        if not attrs:
+            raise ValidationError("Either from_jobs or from_reports is required.")
+        return super().validate(attrs)
+
+
+class ReportOverviewSerializer(ModelSerializer):
+    status = CharField(source="scanjob.status", read_only=True)
```

Comment:
How will that work when report does not have scanjob associated? That's most common when I run a scan and then delete it, but you could also import report using CLI or merge reports from 2 scans (I'm not super confident about these last two cases).

Does fingerprinting create a new scanjob? I thought it does not. So I think we can't put `scanjob.status` value into `status` field here, because report status should be in {fingerprint_not_run, fingerprint_running, fingerprint_complete, ...errors}.

Reply:
It won't, but reports always have a scanjob. Even uploaded or merged reports have a scanjob that contains a single ScanTask with type set to fingerprint. What they lack is a Scan instance.

I've been thinking about what Brad said and maybe we should simply use the scanjob API the part that concern reports API in this PoC. I don't see us moving away from having scanjobs in the near future...